Taper Design Recipe with FDTD and EME

seanlam-ece opened the pull request:
**Overview**
This is a prototype design recipe flow for tapers. Goals include:
1. Showcasing the design recipe flow that can be extended to other plugins
2. Showcasing Lumerical FDTD and EME plugin features
3. Showcasing the silicon photonic taper design recipe for routing

**Design Recipe**
The taper design recipe flow works as follows:
1. Sweep taper geometries and taper lengths in EME. Find the best taper geometry and length that aims to match the equivalent dB/cm loss of routing in the narrow cross section while minimizing reflections.
2. Once best taper component is found in EME, port geometry over to FDTD for s-parameter extraction. Reflections at the ports may be revealed here if abrupt geometries are located at the ports.

**FDTD Plugin Features**
- Write s-parameters - Write / read s-parameters (may run s-parameter simulation to get s-params). Same as previous write_sparameters_lumerical.
- Mesh convergence - Sweeps mesh and monitors s-parameters to minimize variation in s-parameters below a threshold
- Port resizing convergence -  Sweeps E-field intensity thresholds that are maintained at the edges of ports and monitors s-parameters to minimize variation in s-parameters below a threshold. 
- Port resizing - Ensures the edges of the ports decay to a specified E-field intensity. This resizes ports and FDTD simulation region to minimize simulation volume while maintaining simulation accuracy. 


**EME Plugin Features**
- Length sweep - Sweeps center length of device and extracts |S21|^2 and |S11|^2
- Mode coupling coefficients - Extracts mode coupling coefficients along device
- Effective index - Extracts effective index along device
- Mesh convergence - Sweeps mesh and monitors s-parameters to minimize variation in s-parameters below a threshold
- Cell convergence - Sweeps number of cells in center region of device and monitors s-parameters to minimize variation in s-parameters below a threshold
- Mode convergence - Sweeps number of modes considered and monitors s-parameters to minimize variation in s-parameters below a threshold
- Overall convergence (mesh + cell convergence) - Better algorithm that sweeps mesh and number of cells simultaneously to ensure simulation convergence

More info to be added, including diagrams and presentation.

seanlam-ece committed:
[LUM] Convert layerstack to process file

seanlam-ece committed:
[LUM] Replace geometry generation with Layer Builder version

seanlam-ece committed:
[LUM] Handle dopants gracefully for Lum versions 2021 and 2023

seanlam-ece committed:
[LUMERICAL] Add draft EME setup

seanlam-ece committed:
Merge branch 'main' into design-recipe-flow

seanlam-ece committed:
[LUMERICAL] Add base EME simulation creation

seanlam-ece committed:
[LUM] Add EME mesh convergence; fix wavelength units; set group spans based on component geometry

seanlam-ece committed:
[LUM] Add cell convergence

seanlam-ece committed:
[LUM] Add mode convergence testing

seanlam-ece committed:
[LUM] Add length sweep feature

seanlam-ece committed:
[LUM] Return convergence results

seanlam-ece committed:
[LUM] Add pytest for EME

seanlam-ece committed:
[LUM] Switch print to logger

seanlam-ece committed:
[LUM] Use global debug var

seanlam-ece committed:
[LUM] Rename to_lbr function

seanlam-ece committed:
[LUM] Combine setting params into dict for better readability; rename to_lbr

seanlam-ece committed:
[LUM] Update lbr return docstring

seanlam-ece committed:
[LUM] Add flag to enable/disable doping layer generation; update docstrings

seanlam-ece committed:
[LUM] Use global var to enable/disable dopings

seanlam-ece committed:
[LUM] Make global var for enabling/disabling doping

seanlam-ece committed:
[LUM] Loop through FDTD, CHARGE, and MODE to test lbr generation; reformat using black

seanlam-ece committed:
[LUM] Add docstring to describe test

seanlam-ece committed:
Format using black

seanlam-ece committed:
[LUM] Make common geometry import function with more detailed error messages

seanlam-ece committed:
Merge branch 'main' of https://github.com/gdsfactory/gplugins

seanlam-ece committed:
Merge branch 'main' into lumerical-eme

seanlam-ece committed:
[LUM] Use draw_geometry function; fix issue with return data

seanlam-ece committed:
[LUM] Reduce test code

seanlam-ece committed:
[LUM] Clarify comment on dopants causing crashes in Lumerical

seanlam-ece committed:
[LUM] Check each simulator for importing process file

seanlam-ece committed:
Fix issues with pre-commit

seanlam-ece committed:
Merge branch 'main' into lumerical-eme

seanlam-ece committed:
[LUM] Move EME sim and convergence settings to different file

seanlam-ece committed:
[LUM] Move EME test to tests directory

seanlam-ece committed:
Add design recipe base files

simbilod committed:
update pyproject

seanlam-ece committed:
[LUM] Extract mode coupling coefficients; fix pre-commit issues

seanlam-ece committed:
[LUM] Extract neff vs position

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.2.1 → v0.2.2](https://github.com/astral-sh/ruff-pre-commit/compare/v0.2.1...v0.2.2)

seanlam-ece committed:
Add taper design recipe with EME

- Hash layer stack
- Add design recipe eval decorator
- Save convergence results
- Save length sweep
- Ensure lengths are returned as numbers rather than arrays
- Update definition of group span (first group starts from 0)

seanlam-ece committed:
Add taper design recipe

- Add update_overall_convergence that uses both mesh and cell
  convergence to allow for EME sims to converge better
- Update base simulation settings to better converge
- Format classes via pre-commit

seanlam-ece committed:
Skip component if simulation fails to complete

seanlam-ece committed:
[LUM] Move FDTD features to dedicated sim class

seanlam-ece committed:
Ensure taper design recipe sweeps through different geometries

seanlam-ece committed:
[LUM] Add port resizing/convergence test for E-field decay

seanlam-ece committed:
[LUM] Update comments on port convergence

seanlam-ece committed:
[LUM] Add FDTD mesh convergence test; format via pre-commit

seanlam-ece committed:
[LUM] Refactor to extract sparams as dataframe; ensure simulation is saved after convergence.

seanlam-ece committed:
[LUM] Add draft of sweeping field intensity threshold

seanlam-ece committed:
[LUM] Update field intensity threshold for FDTD sims

- Ensure dirpath is casted as pathlib.Path object
- Move field intensity threshold sweep into FDTD sim class for updating
  intensity threshold based on sparam convergence

seanlam-ece committed:
Ensure design recipes handle both components and component generators

seanlam-ece committed:
Add comments to taper design recipe

seanlam-ece committed:
Add hashing to taper design recipe and use bytes rather int for hash

seanlam-ece committed:
Format via pre-commit

seanlam-ece committed:
Ensure __hash__ returns int (other return types not allowed)

seanlam-ece committed:
Ensure simulation setup and convergence setup does not change upon eval to ensure hash remains same

seanlam-ece committed:
[LUM] Separate E-field intensity convergence from updating ports

seanlam-ece committed:
Add draft FDTD design recipe

seanlam-ece committed:
[LUM] Set material fit coefficients in FDTD

seanlam-ece committed:
Remove sample code

seanlam-ece committed:
[LUM] Add port translation away and toward device

dependabot[bot] committed:
Bump release-drafter/release-drafter from 5 to 6

Bumps [release-drafter/release-drafter](https://github.com/release-drafter/release-drafter) from 5 to 6.
- [Release notes](https://github.com/release-drafter/release-drafter/releases)
- [Commits](https://github.com/release-drafter/release-drafter/compare/v5...v6)

---
updated-dependencies:
- dependency-name: release-drafter/release-drafter
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

simbilod committed:
fix xy meshing

joamatab committed:
Merge pull request #343 from simbilod/parametrize_cellname_density

add cellname parametrization, update pyproject for pre-commit to run

joamatab committed:
Merge pull request #348 from gdsfactory/dependabot/github_actions/release-drafter/release-drafter-6

Bump release-drafter/release-drafter from 5 to 6

joamatab committed:
Merge pull request #349 from simbilod/fix_xy_meshing

fix xy meshing

seanlam-ece committed:
Complete taper design recipe

- If sparam simulation fails or sparams cannot be loaded, ensure mesh sweep continues and covers what has run
- Make cross sections as inputs to design recipe. Cross sections cannot be part of design intent since they cannot be hashed.
- Make list of components that have been simulated in EME to summarize results
- Set maximum reflection power to consider. Anything above this threshold will mean that length and reflection will determine the best device. Anything below this threshold will mean that only length will determine the best device.
- Set EME simulation settings lower to observe convergence

seanlam-ece committed:
[EME] Save |S|^2 to csv

seanlam-ece committed:
[LUM] Move convergence files to convergence directory

seanlam-ece committed:
Store recipe runs in recipe_runs directory

seanlam-ece committed:
Save summarized results

seanlam-ece committed:
Update comments and remove test code

seanlam-ece committed:
[LUM] Move material map to simulation settings

seanlam-ece committed:
[LUM] Move constants to config

seanlam-ece committed:
[LUM] Add basic test for FDTD

seanlam-ece committed:
Merge branch 'design-recipe-flow' of https://github.com/elamdf/gplugins into design-recipe-flow

seanlam-ece committed:
Merge branch 'main' of https://github.com/gdsfactory/gplugins

seanlam-ece committed:
Merge branch 'main' into lumerical-fdtd

seanlam-ece committed:
Merge branch 'main' into design-recipe-flow

seanlam-ece committed:
Merge design-recipe-flow into lumerical-fdtd

seanlam-ece committed:
Clean up FDTD recipe

- Reformat via pre-commit
- Move FDTD recipe test to test directory
- Ensure Lumerical.py does not run in __init__.py as it has bug

joamatab committed:
add drc samples

joamatab committed:
install klayout as part of kvis

joamatab committed:
simpler testing

joamatab committed:
simpler installer

seanlam-ece committed:
Rename recipe

joamatab committed:
add klayout

joamatab committed:
Merge pull request #350 from gdsfactory/add_drc_samples

add drc samples

joamatab committed:
Bump to 0.10.0

joamatab committed:
update changelog

joamatab committed:
fix gc prefix

joamatab committed:
Merge pull request #351 from gdsfactory/fix_gc_prefix

fix grating prefix for tidy3d plugin

joamatab committed:
Bump to 0.10.1

joamatab committed:
update changelog

joamatab committed:
Merge pull request #345 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

joamatab committed:
update sax

joamatab committed:
ignore elmer tests

joamatab committed:
fix schematic editor

joamatab committed:
remove elmer tests that are failing

joamatab committed:
comment elmer notebook

joamatab committed:
better names for cicd

joamatab committed:
Merge pull request #354 from gdsfactory/update_sax

Update sax

joamatab committed:
Bump to 0.10.2

joamatab committed:
update changelog

seanlam-ece committed:
Hash design recipes before and after eval; Add FDTD verification to taper design recipe; Update FDTD recipe convergence

elamdf commented:
does "minimize variation in s-parameters" mean "minimize variation [between independent runs]..."?

elamdf commented on gplugins/design_recipe/ConstituentRecipes.py:
```diff
@@ -28,7 +25,7 @@ class ConstituentRecipes:
     iter_lock: threading.Lock
 
-    def __init__(self, recipes: List[dr.DesignRecipe] = []):
+    def __init__(self, recipes: list[dr.DesignRecipe]):
```
a design recipe should always have a non-`None` constituent recipe, right? is the idea to just pass an explicit empty list for that case (the one where a DR has no dependencies)?

elamdf commented on gplugins/design_recipe/ConstituentRecipes.py:
```diff
@@ -38,16 +35,17 @@ def __iter__(self):
-            "you can't concurrently iterate through  multiple instances of this ConstituentRecipes"
+        assert (
+            self.idx_to_yield == 0
+        ), "you can't concurrently iterate through  multiple instances of this ConstituentRecipes"
```
nit: double space "through  multiple"

elamdf commented on gplugins/design_recipe/DesignRecipe.py:
```diff
@@ -25,57 +27,64 @@ class DesignRecipe:
-    component: Optional[Component] = None
+    # The component factory this DesignRecipe operates on. This is not necessarily
+    # the same `component` referred to in the `dependencies` recipes.
+    cell: ComponentFactory | Component | None = None
```
`ComponentFactory` is a callable, right? Would `cell` need to either be evaluated or not evaluated (to get the design) depending on if it's a `ComponentFactory` or a `Component`?

elamdf replied:
Ah I see that logic is handled- curious, why might we want to have `cell` be a `ComponentFactory`? Couldn't we cast it to a `Component` before passing it into a DR?

seanlam-ece replied:
What do you mean by casting? Are you still able to perform a geometry sweep? If not, a ComponentFactory is more applicable since we can sweep its geometry parameters

elamdf commented on gplugins/design_recipe/DesignRecipe.py:
```diff
@@ -25,57 +27,64 @@ class DesignRecipe:
         This is used to determine 'freshness' of a recipe (i.e. if it needs to be rerun)
+
+        Hashed items:
+        - component or cell
```
clarify that it's the component/cell geometry not the whole thing (which might include non-static state)

elamdf commented on gplugins/design_recipe/DesignRecipe.py:
```diff
@@ -25,57 +27,64 @@ class DesignRecipe:
+            recipe.is_fresh() for recipe in self.dependencies
+        )
 
     def eval(self, force_rerun_all=False) -> bool:
```
oops there should be a comment on `force_rerun_all` (it forces this DR and its dependencies to re-`eval()` regardless of freshness)

elamdf commented on gplugins/design_recipe/DesignRecipe.py:
```diff
@@ -25,57 +27,64 @@ class DesignRecipe:
         """
         success = self.eval_dependencies(force_rerun_all)
 
-# TODO find some way to automatically hook into eval()'s subclasses
```
this todo is kind of important- the reason we can't just have implementers of this function in subclasses call `super()` is that in doing that we can't both evaluate dependencies at the beginning and hash at the end. Maybe the hash update thing is a job for a decorator?

elamdf replied:
ah I see below. Nice!

elamdf commented on gplugins/design_recipe/ConstituentRecipes.py:
```diff
@@ -1,25 +1,22 @@
 
-
 # moved into a class in case we want to make this a parallel-dispatch dependency DAG instead of a serial array
 class ConstituentRecipes:
```
not 100% sure that this is the right name (I feel like "constituent" implies that these encompass the functionality of a recipe) but we can roll with it for now, not a major thing

elamdf commented on gplugins/klayout/get_density.py:
```diff
@@ -41,6 +41,7 @@ def put(
 def calculate_density(
     gdspath: Path,
     layer: tuple[int, int],
+    cellname: str | None = None,
```
should cellname be `None` or an empty string? I genuinely don't know, just wary of setting `str`s to `None`

elamdf commented on gplugins/klayout/get_density.py:
```diff
@@ -290,6 +297,7 @@ def estimate_weighted_global_density(
 def plot_density_heatmap(
     gdspath: Path,
     layer: Layer,
+    cellname: str | None = None,
```
same q about default value

elamdf commented on gplugins/lumerical/Lumerical.py:
```diff
@@ -1,21 +1,26 @@
+import numpy as np
+import yaml
+from gdsfactory.config import PathType, logger
+from gdsfactory.generic_tech.simulation_settings import (  # TODO wildcard imports are bad
```
remove TODO

elamdf commented on gplugins/lumerical/Lumerical.py:
```diff
@@ -1,21 +1,26 @@
 )
-from gdsfactory.typings import Component, LayerStack, MaterialSpec
-from gdsfactory.config import PathType, __version__, logger
-from gdsfactory.generic_tech.simulation_settings import SimulationSettingsLumerical, SimulationSettingsLumericalFdtd, SIMULATION_SETTINGS_LUMERICAL_FDTD # TODO wildcard imports are bad
```
remove TODO

elamdf commented on gplugins/lumerical/Lumerical.py:
```diff
@@ -34,32 +39,32 @@ class Engine(Enum):
     CHARGE = 1
```
this is unused

elamdf commented on gplugins/lumerical/Lumerical.py:
```diff
@@ -1,21 +1,26 @@
+)
 from gplugins.lumerical.write_sparameters_lumerical import set_material
 
 DEBUG_LUMERICAL = 1
```
comment? "This enables/disables displaying the Lumerical GUI when interacting with sessions" or something

elamdf commented on gplugins/lumerical/Lumerical.py:
```diff
@@ -134,24 +142,29 @@ def draw_geometry(component: Component,
+    for (
+        name,
+        layer_obj,
+    ) in layer_stack.layers.items():  # layer, thickness in layer_to_thickness.items():
```
remove comment

elamdf commented on gplugins/lumerical/Lumerical.py:
```diff
@@ -370,11 +402,12 @@ def __init__(self,
         pass
-        #write_charge_distribution_lumerical(self.component,
-
+        # write_charge_distribution_lumerical(self.component,
```
do we want this line?

elamdf commented on gplugins/lumerical/config.py:
```diff
@@ -0,0 +1,24 @@
+DEBUG_LUMERICAL = False
```
this conflicts with `DEBUG_LUMERICAL` in `Lumerical.py`

elamdf replied:
should probably just remove that one and keep things in `CONFIG`

elamdf replied:
still needs a comment though

elamdf commented on gplugins/lumerical/convergence_settings.py:
```diff
@@ -0,0 +1,39 @@
+    Lumerical EME convergence settings
+
+    Parameters:
+        passes: Number of passes / simulations sweeping a convergence parameter before checking for convergence
```
what does this mean? Is this a performance optimization (only check for convergence after running x sweeps so as to not slow them down)?

elamdf replied:
or is this the number of runs we consider gauge variance/consistency?

seanlam-ece replied:
Number of runs we consider for variance. Ex. If passes = 4, we run sparam extraction for mesh_cells_per_wavelength = [0,1,2,3,4,5,6,7,8,9,10] and we get sparams = [data0, data1, data2, data3, ... data10]. We take the max diff in sparams from:
sparams[4] - sparams[0]
sparams[4] - sparams[1] 
sparams[4] - sparams[2] 
sparams[4] - sparams[3]

to check for convergence.

elamdf commented on gplugins/lumerical/convergence_settings.py:
```diff
@@ -0,0 +1,39 @@
+
+    Parameters:
+        passes: Number of passes / simulations sweeping a convergence parameter before checking for convergence
+        sparam_diff: Maximum difference in sparams after x passes sweeping a convergence parameter. Used to check for convergence.
```
this should be a percentage tolerance, right?

seanlam-ece replied:
Change convergence tolerance to |S_n - S_o| / S_o to use relative convergence.

Update: Absolute convergence is still better than relative convergence since computational / quantization noise prevents relative convergence from completing.

seanlam-ece replied:
Need to discuss further since relative convergence also has issues.

seanlam-ece replied:
Settled on absolute convergence since quantization / computation noise prevents relative convergence from completing

elamdf commented on gplugins/lumerical/convergence_settings.py:
```diff
@@ -0,0 +1,39 @@
+
+    Parameters:
+        port_field_intensity_threshold: E-field intensity at the edge of each port. Used to resize ports and FDTD region.
+        sparam_diff: Maximum difference in sparams after x passes sweeping a convergence parameter. Used to check for convergence.
```
same comment about % vs abs tolerance

seanlam-ece replied:
Settled on absolute convergence since quantization / computation noise prevents relative convergence from completing

elamdf commented on gplugins/lumerical/convergence_settings.py:
```diff
@@ -0,0 +1,39 @@
+    Lumerical FDTD convergence settings
+
+    Parameters:
+        port_field_intensity_threshold: E-field intensity at the edge of each port. Used to resize ports and FDTD region.
```
what does this mean? what are the units/is this a normalized term? is this 'desired E-field intensity'?

seanlam-ece replied:
Units: (V/m)^2 
The E-field intensity profile is normalized such that the highest E-field intensity in the mode profile is 1.
Yes, it is the desired E-field intensity at the edges of ports.

elamdf commented on gplugins/lumerical/convergence_settings.py:
```diff
@@ -0,0 +1,39 @@
+    port_field_intensity_threshold: float = 1e-5
+
+    class Config:
+        arbitrary_types_allowed = True
```
why is this necessary?

seanlam-ece replied:
Ports launch a particular mode into the FDTD simulation region. These ports need to be sized properly to ensure the an accurate mode is launched.

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+from gplugins.lumerical.utils import draw_geometry, layerstack_to_lbr
+
+
+def main():
```
this looks like a demo- maybe it should be in an `examples` area somewhere?

seanlam-ece replied:
Agreed

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+
+class LumericalEmeSimulation:
+    """
+    Lumerical EME simulation
```
maybe "A plugin for performing Lumerical EME simulations on GDSFactory designs"

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        **settings,
+    ):
+        """
+        Set up EME simulation based on component geometry and simulation settings. Optionally, run convergence.
```
"convergence testing/optimization"?

seanlam-ece replied:
No optimization in here. This file should only include methods that assist with running simulations.

elamdf replied:
ah I meant convergence optimization- just 'run convergence' is a little unclear

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        Set up EME simulation based on component geometry and simulation settings. Optionally, run convergence.
+
+        Parameters:
+            component: Component geometry to simulate
```
"The `Component` to simulate"

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        simulation_settings.update(**settings)
+        ss = SimulationSettingsLumericalEme(**simulation_settings)
+
+        # Check number of cell groups are aligned
```
this should probably go after session initialization? broadly, this is a very long function that maybe could be broken up a little?

elamdf replied:
drawing component geometry (ie generating the extensions/boolean instances and calling draw_geo) should definitely be its own method, setting up the simulating settings as well probably

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        convergence_data = pd.DataFrame.from_dict(
+            {"mesh_cells": mesh_cells_per_wavl, "s21": list(s21), "s11": list(s11)}
+        )
+        convergence_data.to_csv(
```
maybe we should have a `save()` method that takes an enum or something for the kind of data? thinking about how we can make this as easy as possible to map to a database (rather than individual files) later

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        Updates both Lumerical session and simulation settings.
+
+        Parameters:
+            plot: Plot and save convergence results
```
"Whether to plot and.."

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+            num_cells.append(ss.group_cells[1:-1])
+
+            for i in range(1, len(ss.group_cells) - 1):
+                ss.group_cells[i] += 1
```
I don't see where simulation parameters are being updated to optimize for convergence- is it here (if so, is all this does increase fidelity)? should probably have a comment

seanlam-ece replied:
Line 418 is where it updates the simulation parameter. ss.group_cells[i] += 1 increases the number of cells (fidelity). Will add a comment.

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+
+    def update_cell_convergence(self, plot: bool = False) -> pd.DataFrame:
+        """
+        Update simulation based on cell convergence testing (number of slices across the device center).
```
This ensures (or attempts to ensure) convergence, maybe should be named accordingly? let's chat about this at 10

elamdf replied:
still kind of unsure how it does that

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        )
+        return convergence_data
+
+    def update_mode_convergence(self, plot: bool = False) -> pd.DataFrame:
```
are these `update_*_convergence` different convergence parameters that are swept? This seems like not the right way to do a multi-axis sweep (especially when there might be interdependencies on convergence). we can discuss

seanlam-ece replied:
As discussed, I started with a single simulation parameter convergence testing. Multi-dimensional convergence testing is better, but difficult to execute. I provide a method called "update_overall_convergence", which from my testing, combines the cell and mesh convergence to achieve better convergence capabilities. Mode convergence is in its separate domain since it performs well on its own and the main culprits to simulation inaccuracy come from mesh and cells.

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        group: int = 1,
+    ) -> pd.DataFrame:
+        """
+        Get length sweep sparams.
```
what is a length sweep? is this specific to the taper example or is it generically applicable to EME components?

seanlam-ece replied:
Generically applicable to EME components. EME simulated components are two port, long structures, such as tapers, mode converters, interlayer transition devices, etc.

elamdf replied:
this is a sweep of the simulation group span, right? might want to specify that we aren't actually sweeping the length of the device (unless I'm mistaken and we are, in which case clarify that instead)

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+            | float | float | ...
+            3. overlap - Overlap integrals (dB)
+            | mode1 | mode2 | ...
+            | float | float | ...
```
these are arrays of floats, right?

seanlam-ece replied:
Pandas DataFrames of floats

elamdf replied:
sure, should still specify that it's not a scalar float

seanlam-ece replied:
It is a column of scalar floats. If each entry has an array, it would be specified like:
https://github.com/elamdf/gplugins/pull/3#discussion_r1518384056

Thoughts on documenting this way?

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        ss = self.simulation_settings
+        max_coupled_mode = max_coupled_mode or ss.num_modes
+
+        if s.layoutmode():
```
prob comment "If the simulation hasn't been run, run it." or something

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        # Get starting point for number of cell and group span
+        cells = [int(val[0]) for val in s.getnamed("EME", "cells")]
+        num_cells = cells[group]
+        group_span = round(s.getnamed("EME", "group spans")[group][0], 15)
```
is 15 a magic number for a quantity in lumerical?

seanlam-ece replied:
No, this is to cover floats converted from Lumerical to Python that have limited precision. Ex. if in Lumerical, it may get a group span as 1um, but in Python it may be casted to 0.9999999um.

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+            S = s.getresult(f"EME::Cells::cell_{i}", f"S_{i}_{i+1}")
+
+            # Get coupling based on S parameters and overlap coefficients
+            if i == cell_start + 1:
```
you can just initialize `S_bcoupling` to be an empty array, right? since you're concatenating

seanlam-ece replied:
If I initialized `S_bcoupling` as an empty array, it has to:
1. Have a defined shape
2. Filled with default values
3. Then be replaced with the first set of data

Perhaps there is a cleaner way, but I have settled on this method.

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+
+            # Get coupling based on S parameters and overlap coefficients
+            if i == cell_start + 1:
+                # Create initial column
```
same comment as above re: initialization to avoid this duplicated logic

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        position = np.linspace(0, group_span, num_cells + 1)[1:]
+
+        coupling_coefficients = {
+            "forward_sparam": pd.DataFrame.from_dict(
```
you can probably do something like
```
"forward_sparam": pd.DataFrame({f"mode{i+1}": S_fcoupling[i] for i in range(S_fcoupling.shape[0])})
```
to make this more concise (and probably faster)

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+        self, input_mode: int, max_coupled_mode: int = None, group: int = 1
+    ) -> None:
+        """
+        Plot mode coupling coefficients from the input_mode to the max_coupled_mode.
```
plotted against position, right?

seanlam-ece replied:
Yes, I'll update the comment

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+
+    def get_neff_vs_position(self, group: int = 1) -> pd.DataFrame:
+        """
+        Get effective index vs position along device
```
"...along device for a given cell group"

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+            1. Effective index for different modes
+            2. Position (m)
+            | position | mode1 | mode2 | ...
+            | float    | float | float | ...
```
`array(float)` or something

seanlam-ece replied:
I assume the columns are already considered arrays. I denote float meaning the entire column has data of type floats. 

Other pandas dataframes may have np.arrays as the data in the dataframe. Ex. multi-wavelength sparams:

```
| mesh_accuracy | wavelength       | S11        | S12 ...
| int           | np.array         | np.array   | np.array ...
| 1             | [1.5e-6, 1.6e-6] | [1, 0.2]   | [0.2, 1]
```

elamdf commented on gplugins/lumerical/eme.py:
```diff
@@ -0,0 +1,1063 @@
+            s21.append(abs(S[1, 0]) ** 2)
+            mesh_cells_per_wavl.append(ss.mesh_cells_per_wavelength)
+
+            ss.mesh_cells_per_wavelength += 1
```
should probably have a comment here

seanlam-ece replied:
Done in latest
https://github.com/seanlam97/gplugins/blob/a135b4bfcfbe6db972d8e3297c5291c593369a0d/gplugins/lumerical/eme.py

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+
+run=False returns the simulation session for you to debug and make sure it is correct
+
+To compute the Sparameters you need to pass run=True
```
"
WARNING: You have passed run=False to `LumercialFdtdSimulation`!

run=False returns the un-evaluated simulation session for debugging purposes.

To compute S-parameters, pass run=True"

seanlam-ece replied:
This seems to be legacy and can be removed.

seanlam-ece replied:
Removed in latest:
https://github.com/seanlam97/gplugins/blob/a135b4bfcfbe6db972d8e3297c5291c593369a0d/gplugins/lumerical/fdtd.py

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+"""
+
+
+def main():
```
examples should go outside of the library file

seanlam-ece replied:
Moved to examples:
4fe395fa77f5e5f571b9f308771cabcc2138fb9c

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+    """
+    Lumerical FDTD simulation
+
+    Set up FDTD simulation based on component geometry and simulation settings. Optionally, run convergence.
```
clarify what it means to "run convergence"- "run converge optimization sweeps"?

seanlam-ece replied:
Done: 37a80b88156ec79b05cc6366dd29d05acd5f59ec

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+        dirpath: Directory where simulation files are saved
+        filepath_npz: S-parameter filepath (npz)
+        filepath_fsp: FDTD simulation filepath (fsp)
+        mesh_convergence_data: Mesh convergence results
```
is this passed from a function? should specify

seanlam-ece replied:
Passed from `update_mesh_convergence` method. Will update comment

seanlam-ece replied:
Done in latest:
https://github.com/seanlam97/gplugins/blob/a135b4bfcfbe6db972d8e3297c5291c593369a0d/gplugins/lumerical/fdtd.py

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+        simulation_settings: FDTD simulation settings
+        convergence_settings: FDTD convergence settings
+        dirpath: Directory where simulation files are saved
+        filepath_npz: S-parameter filepath (npz)
```
these are output filepaths, right? should specify

seanlam-ece replied:
Yes

seanlam-ece replied:
Done in latest:
https://github.com/seanlam97/gplugins/blob/a135b4bfcfbe6db972d8e3297c5291c593369a0d/gplugins/lumerical/fdtd.py

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+
+        Your components need to have ports, that will extend over the PML.
+
+        .. image:: https://i.imgur.com/dHAzZRw.png
```
what does this show? the ports extending beyond PML?

seanlam-ece replied:
This is legacy, but this shows an example of how the simulation region will be generated.

elamdf replied:
does this lead to an autogenerated doc or something that's more clear about what this shows?

seanlam-ece replied:
Yes, I see this in my PyCharm IDE. Other IDEs may compile this image similarly.

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+        - dirpath
+        - layerStack
+
+        converts gdsfactory units (um) to Lumerical units (m)
```
is this just something that the function does? not sure why this is here/where this happens

seanlam-ece replied:
The function does this. GDSFactory numbers are multiplied by a constant called `um` to convert to Lumerical. This constant is placed in gplugins/lumerical/config.py

elamdf replied:
can you add a comment that says that?

seanlam-ece replied:
Done: 37a80b88156ec79b05cc6366dd29d05acd5f59ec

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+            field_profile_samples: number of wavelengths to compute field profile.
+
+
+        .. code::
```
this is extended with oxide, right? should maybe clarify?

seanlam-ece replied:
Layerbuilder handles the background material via the process file (.lbr)

elamdf commented on gplugins/lumerical/fdtd.py:
```diff
@@ -0,0 +1,1282 @@
+        Returns:
+            S-parameters vs wavelength
+            | wavelength | S11     | S12     | ...
+            | float      | complex | complex | ...
```
array<complex>?

seanlam-ece replied:
See and discuss here: https://github.com/elamdf/gplugins/pull/3#discussion_r1518073342

seanlam-ece committed:
Add rough draft of results saving

joamatab committed:
add models

joamatab committed:
increase number of errors

joamatab committed:
remove post_process

joamatab committed:
fix sdl

elamdf commented on gplugins/lumerical/recipes/fdtd_recipe.py:
```diff
@@ -0,0 +1,204 @@
+        """
+        h = hashlib.sha1()
+        int_hash = super().__hash__()
+        h.update(int_hash.to_bytes(int_hash.bit_length() + 7 // 8, byteorder="big"))
```
7 // 8 will always return 0- is this intended?

seanlam-ece replied:
Good catch, this was a mistake

elamdf commented on gplugins/lumerical/recipes/fdtd_recipe.py:
```diff
@@ -0,0 +1,204 @@
+)
+
+
+def example_run_fdtd_recipe():
```
maybe better to have this in an examples file/dir given it's like half of the total file length

seanlam-ece replied:
Done:
82f6483c7ab588d534fba4a96118158e28dbd3e3

elamdf commented on gplugins/lumerical/recipes/taper_design_recipe.py:
```diff
@@ -0,0 +1,389 @@
+)
+
+
+def example_run_taper_design_recipe():
```
Great job getting this set up! One thought- `DesignRecipe`s were designed to be composed- maybe this could be refactored as an EME design recipe, an FDTD design recipe, and a `RoutingTaperDesignRecipe` with the following structure

`RoutingTaperDesignRecipe` depends on `FDTDDesignRecipe` which depends on `EMEDesignRecipe`.

Then in the example you can just do `RoutingTaperDesignRecipe.eval()` and it'll evaluate all the dependencies automatically.

seanlam-ece replied:
The notion of an EMERecipe is tough to define since EME can be used in different ways (length sweeps, modal analysis, neff engineering) and applications (tapers, bragg gratings, couplers, bends).

The issue here is that specific data from the EME sims need to be passed onto the FDTD recipe before eval'ing. Data seems to be easiest to be passed in the same eval method from one recipe to another. Any other ideas on this?

elamdf commented on gplugins/lumerical/tests/test_fdtd_recipe.py:
```diff
@@ -1,18 +1,16 @@
     c = gf.Component("two_references")
     wr1 = c << gf.components.straight(width=0.6, layer=LAYER.WG)
     wr2 = c << gf.components.straight(width=0.6, layer=LAYER.WG)
     wr2.movey(10)
```
might be good to test the `is_fresh()` of these too (create one with a component, run it, check `is_fresh() == True` and then change the component and check `is_fresh() == False`)

seanlam-ece replied:
Done: e9cc257130bf07d26894e29cfb6a058dba281d8d

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,267 @@
+        "Aluminum": "Al (Aluminium) Palik",
+    }
+
+    # Create layerstack
```
is this common? can it be moved to a different file and imported? just so long

seanlam-ece replied:
Removed as much unnecessary code as possible. Layerstack definition is custom here.

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,267 @@
+            + "and subsequent versions of Lumerical that support both passives and dopants."
+        )
+    logger.info(message)
+    assert success, message
```
ideally we have some functional verification (import component + layerstack, run sim and make sure it converges or something) might be able to have some reuse with the corresponding designrecipe tests

seanlam-ece replied:
Yes, this is done via test_eme.py and test_fdtd.py. Here, the purpose is a unit test for importing process file since the process file creation is reversed engineered from the output of Lumerical's software. If the process file format changes in the future, we can catch it here.

elamdf commented on gplugins/lumerical/write_charge_distribution_lumerical.py:
```diff
@@ -295,12 +305,12 @@ def write_charge_distribution_lumerical(
     add_charge_materials(s)
 
-    # add si and 
-    print( component_with_booleans.get_layers())
```
should use gdsfactory logging lib?

elamdf reviewed (changes requested):
Great work on this! Left some comments but overall it's shaping up well :)

joamatab committed:
Merge pull request #357 from gdsfactory/add_models

Add models

seanlam-ece committed:
[LUM] Add draft DEVICE simulation class

seanlam-ece committed:
[LUM] Set CHARGE simulation region

seanlam-ece committed:
Update comments and reformat

seanlam-ece committed:
Clean up default vals for recipes; remove redundant hashing calculation

seanlam-ece committed:
Fix is_fresh check

seanlam-ece committed:
Add comments about port field intensity

seanlam-ece committed:
Move example FDTD recipe to separate file+directory

seanlam-ece committed:
Ensure convergence directory created before saving files there; add test for recipe freshness

seanlam-ece committed:
Ensure EME portion of taper design recipe returns optimal component lengths for each respective geometry

seanlam-ece committed:
Remove default settings

seanlam-ece reviewed (commented):
I implemented changes based on your feedback and left a few comments. The Lumerical.py and write_charge_distribution_lumerical.py are to be replaced by fdtd.py and device.py. DEVICE functionality will be in a different PR.

seanlam-ece committed:
Add comment

seanlam-ece committed:
Move example code to examples dir; Add comments

seanlam-ece committed:
Reformat via pre-commit

seanlam-ece committed:
Add draft of result saving

seanlam-ece committed:
[LUM] Verify result saving and recall to skip convergence testing

seanlam-ece committed:
Format via pre-commit

seanlam-ece committed:
Add comments about FDTD convergence results

seanlam-ece committed:
Save FDTD convergence results only when all convergence tests are done

seanlam-ece committed:
[LUM] Add convergence results recall to skip simulations

seanlam-ece committed:
Format via precommit

seanlam-ece committed:
[LUM] Handle hash collisions by checking setups

seanlam-ece committed:
Format via pre-commit

seanlam-ece committed:
Add comment on convergence results

seanlam-ece committed:
[LUM] Recall convergence results for FDTD sims

- Enables skipping convergence simulations if convergence results are
  available. If available, simulation settings are gleaned from file and
  are set.
- Move port field intensity threshold from convergence settings to simulation settings. This was affecting convergence results recalling since this setting changes after convergence testing.

seanlam-ece committed:
[LUM] Group simulation files into directory labeled with class name and hash

seanlam-ece committed:
Clarify FDTD comments

seanlam-ece committed:
[LUM] Set CHARGE simulation region, settings, and boundary conditions

seanlam-ece committed:
[LUM] Ensure simulation creation starts fresh; label electrical boundaries

seanlam-ece committed:
Merge lumerical-results into lumerical-charge

seanlam-ece committed:
[LUM] Set boundary condition settings; Add charge monitor

seanlam-ece committed:
[LUM] Save design recipe results for recall; Fix up taper design recipe

seanlam-ece committed:
Merge branch 'lumerical-results' into lumerical-charge

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.2.2 → v0.3.3](https://github.com/astral-sh/ruff-pre-commit/compare/v0.2.2...v0.3.3)
- [github.com/PyCQA/bandit: 1.7.7 → 1.7.8](https://github.com/PyCQA/bandit/compare/1.7.7...1.7.8)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

simbilod committed:
fix polygon hole not meshing

seanlam-ece committed:
[LUM] Convert units to m for charge monitor; Partition after all simulation objects created

seanlam-ece committed:
[LUM] Add ability to change+track boundary condition name and settings

seanlam-ece committed:
[LUM] Handle boundary condition names; specify charge profile path"

- Raise exception when attempting to swap names
- Special symbols such as +, -, _ are removed in names

seanlam-ece committed:
[LUM] Fix material name

seanlam-ece committed:
Format via pre-commit

seanlam-ece committed:
Use simulation dirpath to save files

seanlam-ece committed:
[LUM] Add MODE plugin

- Extract index profile, neff, ng, and dispersion and plot
- Perform mesh and port E-field intensity convergence testing
- Save and recall convergence results
- Use example of straight waveguide

seanlam-ece committed:
[LUM] Add comment about changing bc names

seanlam-ece committed:
[LUM] Add simulation and convergence settings for MODE

seanlam-ece committed:
[LUM] Add comments for MODE convergence settings

seanlam-ece committed:
Add test for design recipes; Add dependencies txt file for traceability

seanlam-ece committed:
Merge branch 'lumerical-results' into lumerical-charge

seanlam-ece committed:
Merge branch 'lumerical-results' into lumerical-mode

joamatab committed:
Merge pull request #362 from simbilod/fix_mesh_holes

fix polygon hole not meshing

joamatab committed:
Merge pull request #358 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

seanlam-ece committed:
Simplify FDTD test component

joamatab committed:
add more sax models

joamatab committed:
fix klayout dataprep docs issue

joamatab committed:
fix docs

joamatab committed:
fix docs

joamatab committed:
Merge pull request #364 from gdsfactory/more_models

add more sax models

joamatab committed:
Bump to 0.11.0

joamatab committed:
update changelog

seanlam-ece committed:
Initialize Results instance in __init__ method

- Initializing Results instance as a class variable is problematic.
  Multiple instances of the DesignRecipe will have the same Results
instance, causing it to be overwritten by the latest recipe.

seanlam-ece committed:
Add weighted decision matrix for taper design recipe; Place EME and FDTD recipes into one class for taper design

seanlam-ece committed:
Merge branch 'lumerical-results' of https://github.com/seanlam97/gplugins into lumerical-results

seanlam-ece committed:
Merge branch 'main' of https://github.com/gdsfactory/gplugins

seanlam-ece committed:
Remove pytesting for Lumerical

seanlam-ece committed:
Move lumapi imports to functions

- This ensures that importing lumerical module and files do not
  immediately trigger an exception. Automated doc compilation can run.

seanlam-ece committed:
Catch lumapi exception

seanlam-ece committed:
Merge lumerical-results into lumerical-fdtd

seanlam-ece committed:
[LUM] Move lumapi imports to functions

seanlam-ece committed:
Use absolute path for saving to files. Pickle relative paths

seanlam-ece committed:
Handle OS-dependent paths during unpickling

seanlam-ece committed:
Merge lumerical-results into lumerical-results

seanlam-ece committed:
[LUM] Move constants to config

seanlam-ece committed:
[LUM] Remove old CHARGE and FDTD creation

seanlam-ece committed:
[LUM] Add test for DEVICE

seanlam-ece committed:
Merge lumerical-fdtd into lumerical-charge

seanlam-ece committed:
Move CHARGE example to examples directory

seanlam-ece committed:
Remove boiler plate code

seanlam-ece committed:
Move lumapi import to function (enables automated testing)

seanlam-ece committed:
Move taper design recipe example to examples directory

seanlam-ece committed:
Merge lumerical-results into lumerical-charge

seanlam-ece committed:
Fix bugs in taper design recipe

- Inherit Simulation class for EME simulations
- Log hashed directory and pickled results
- Save length sweeps in taper design recipe rather than single length sweep
- Fix generated data

seanlam-ece committed:
Overwrite Results dirpath if different from specified location

seanlam-ece committed:
Merge branch 'lumerical-mode' into lumerical-results

seanlam-ece committed:
[LUM] Move lumapi import to function

seanlam-ece committed:
[LUM] Create MODE test

seanlam-ece committed:
[LUM] Add test directory for MODE sims

seanlam-ece committed:
Merge main into lumerical-results
